src/publish/index.coffee

```diff
@@ -13,85 +14,79 @@ test = require './test'
 prompt = require './prompt'
 isPrivate = require './public'
 
-EndIf = (log) ->
-  endIf = (exitCodeOrError, message) ->
-    return unless exitCodeOrError?
-
-    if exitCodeOrError instanceof Error
-      log.error exitCodeOrError.message
-      process.exit(1)
-    else
-      return if exitCodeOrError == 0
-      message ?= "exited with #{exitCodeOrError}"
-      log.error message
-      process.exit(exitCodeOrError)
 
 module.exports = (dir, log, config, version, testCommand) ->
   debug "start"
 
-  endIf = EndIf(log)
   git = Git(dir)
   npm = Npm(dir, log)
   changelog = Changelog(dir, git)
 
   if isPrivate(dir)
-    console.log 'Cannot publish this package because it is private.'
+    log 'cannot publish this package because it is private.'
     process.exit(2)
 
-  verify dir, (error) ->
-    endIf(error)
-
-    prep(dir, log, config)
-    debug 'ensured license headers'
+  async.waterfall [
+    (done) ->
+      verify dir, done
 
-    verify dir, (error) ->
-      endIf(error)
+    (done) ->
+      prep(dir, log, config)
+      debug 'ensured license headers'
+      done()
 
-      npm.install config.registry, (error) ->
-        endIf(error)
+    (done) ->
+      verify dir, done
 
-        test dir, log, npm, testCommand, (error) ->
-          endIf(error)
+    (done) ->
+      npm.install config.registry, done
 
-          changelog.build version, (error, tempChangelog) ->
-            endIf(error)
+    (done) ->
+      test dir, log, npm, testCommand, done
 
-            git.getSha (error, sha) ->
-              endIf(error)
+    (done) ->
+      changelog.build version, done
 
-              tempChangelogPath = "/tmp/npub/changelog-#{sha}.md"
-              changelog.write(tempChangelog, tempChangelogPath)
+    (tempChangelog, done) ->
+      git.getSha (error, sha) ->
+        done(error, tempChangelog, sha)
 
-              openEditor tempChangelogPath, (error) ->
-                if error?
-                  fs.unlinkSync tempChangelogPath
-                  endIf(error)
+    (tempChangelog, sha, done) ->
+      tempChangelogPath = "/tmp/npub/changelog-#{sha}.md"
+      changelog.write(tempChangelog, tempChangelogPath)
+      done(null, tempChangeLog)
 
-                changelog.update(tempChangelogPath)
+    (tempChangelogPath, done) ->
+      openEditor tempChangelogPath, (error) ->
+        if error?
+          fs.unlinkSync tempChangelogPath
+        changelog.update(tempChangelogPath)
```

Comment:
Is there a reason to call `changelog.update` when the editor failed? Same for `updateVersion`?